core-bundle/src/Intl/Countries.php

```diff
@@ -0,0 +1,164 @@
+     */
+    private $requestStack;
+
+    /**
+     * @var ContaoFramework
+     */
+    private $contaoFramework;
+
+    /**
+     * @var array<string>
+     */
+    private $countriesList;
+
+    public function __construct(TranslatorInterface $translator, RequestStack $requestStack, ContaoFramework $contaoFramework, array $countriesList)
+    {
+        $this->translator = $translator;
+        $this->requestStack = $requestStack;
+        $this->contaoFramework = $contaoFramework;
+        $this->countriesList = $countriesList;
+    }
+
+    /**
+     * @return array<string,string> Translated country names indexed by their uppercase ISO 3166-1 alpha-2 code
+     */
+    public function getCountries(string $displayLocale = null): array
+    {
+        if (null === $displayLocale && null !== ($request = $this->requestStack->getCurrentRequest())) {
+            $displayLocale = $request->getLocale();
+        }
+
+        $countries = SymfonyCountries::getNames($displayLocale);
+        $needsResort = false;
+
+        if ($this->countriesList) {
+            $countries = $this->filterCountries($countries, $displayLocale ?? 'en');
+            $needsResort = true;
+        }
+
+        foreach (array_keys($countries) as $countryCode) {
+            $langKey = 'CNT.'.strtolower($countryCode);
+
+            if (
+                $langKey !== ($label = $this->translator->trans($langKey, [], 'contao_countries', $displayLocale))
+                && \is_string($label)
+                && '' !== $label
+            ) {
+                $countries[$countryCode] = $label;
+                $needsResort = true;
+            }
+        }
+
+        if ($needsResort) {
+            (new \Collator($displayLocale ?? 'en'))->asort($countries);
+        }
+
+        if (!empty($GLOBALS['TL_HOOKS']['getCountries'])) {
+            return $this->applyLegacyHook($countries);
+        }
+
+        return $countries;
+    }
+
+    /**
+     * @return array<string> Uppercase ISO 3166-1 alpha-2 codes
+     */
+    public function getCountryCodes(): array
+    {
+        // If the legacy hook is used, it might add or remove countries
+        if (!empty($GLOBALS['TL_HOOKS']['getCountries'])) {
+            $countryCodes = array_keys($this->getCountries());
+            sort($countryCodes);
+
+            return $countryCodes;
+        }
+
+        $countryCodes = SymfonyCountries::getCountryCodes();
+
+        if ($this->countriesList) {
+            $countryCodes = array_keys($this->filterCountries(array_combine($countryCodes, $countryCodes), 'en'));
+            sort($countryCodes);
+        }
+
+        return $countryCodes;
+    }
+
+    private function filterCountries(array $countries, string $displayLocale): array
+    {
+        $newList = array_filter(
+            $this->countriesList,
+            static function ($country) {
+                return !\in_array($country[0], ['-', '+'], true);
+            }
+        );
+
+        if ($newList) {
```

Comment:
should this be `!empty()` or `count($newList) > 0`?

Reply:
no, see https://github.com/contao/contao/pull/3138#discussion_r667317566